### PR TITLE
Check the Self-type of inherent associated constants

### PR DESCRIPTION
**Pull request by matthewjasper**
r? @arielb1

### CODE REVIEW

File: src/librustc_mir/hair/pattern/mod.rs [hunk 1]
```diff
@@ -414,61 +403,92 @@ impl<'a, 'tcx> PatternContext<'a, 'tcx> {
...
-                                    err.note("When matching against a range, the compiler \
-                                              verifies that the range is non-empty. Range \
-                                              patterns include both end-points, so this is \
-                                              equivalent to requiring the start of the range \
-                                              to be less than or equal to the end of the range.");
-                                }
-                                err.emit();
-                                PatternKind::Wild
-                            }
+                let lo_lit = self.lower_lit(lo_expr);
```
Comment (arielb1) [inline src/librustc_mir/hair/pattern/mod.rs]: I think a comment would be welcome here

**Comment by matthewjasper** [comment]
Removed the overlap with #58371

File: src/librustc_typeck/check/mod.rs [hunk 2]
```diff
@@ -2236,7 +2236,9 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             hir_id, def_id, substs, user_self_ty, self.tag(),
         );
 
-        if !substs.is_noop() {
+        if !substs.is_noop() || user_self_ty.map_or(false, |ty| {
+            ty.has_free_regions() || ty.has_projections() || ty.has_infer_types()
```
Comment (arielb1) [inline src/librustc_typeck/check/mod.rs]: Is there some justification for why these checks are enough?

Comment (arielb1) [reply src/librustc_typeck/check/mod.rs]: Even if these checks are enough, shouldn't they be run on the pair `(substs, ty)`? i.e.

```rust
let types = (substs, ty); // yes there are TypeFoldable impls for pairs and options
if types.has_free_regions() || types.has_projections() || types.has_infer_types() {
    // ...
}
```

Comment (arielb1) [reply src/librustc_typeck/check/mod.rs]: Nah I've seen a comment at https://github.com/rust-lang/rust/blob/a7ae7a21624952f5bfc6aaf88e53351385b89dd4/src/librustc_typeck/check/mod.rs#L2436-L2444

Maybe move the check there it to a separate function (`needs_user_type_annotation`), while still doing the `types` trick?

**Comment by arielb1** [comment]
r=me with the `needs_user_type_annotation` function

**Pull request merged**